### DISCUSSION TITLE
CRD and controller scope misunderstanding

### DISCUSSION THREAD

**Discussion by dyens** [kubernetes-sigs/kubebuilder#3155] [Support] [answered]
Hi

If i run default `kubebuilder` commands 

```sh
kubebuilder init --domain mydomain --repo mydomain/my-controller
kubebuilder create api --group ai --version v1 --kind MyController
```

I get CRD with `namespaced`  scope, but operator is `cluster` scoped (it listen all namespaces and have cluster role and role-binding specifications). This i a little bit strange for me. What the reason for this behaviour ?

So if i want to use `namespaced` scope operators i need to modify role, role-bindings specifications + pass namespace name to controller manager using environment variables.

My plan is correct?

The second question is:  If i have a lot of namespaces and use my `Namespaced` operator in each of them is it better versus one `Cluster` scope operator (from the performance point of view) ? (in my mind we register many watchers for kubernets events and maybe it can be a problem. On the other hand, if we have a lot of events / instances  - maybe this is a way to paralleling `reconcile` process for different namespaces)

**Comment by camilamacedo86** [comment, marked as answer]
The Operator can be cluster or namespace scoped as the CRD
That means available for all cluster or just in the namespace
For further information check the doc: 
- Operator Scope: https://sdk.operatorframework.io/docs/building-operators/golang/operator-scope/
- CRD scope: https://sdk.operatorframework.io/docs/building-operators/golang/crds-scope/

If your operator is cluster scoped then it will be watching the whole cluster
You can filter the cache with https://github.com/kubernetes-sigs/controller-runtime/blob/main/designs/use-selectors-at-cache.md

However, if you are watching a lot in a big cluster with a lot of things you will probably have performance issues. 

I hope that answer your question
So, I am closing this one. 

however, feel free to re-open if you need.
